### PR TITLE
Workspace trust - empty workspace

### PR DESCRIPTION
**Pull request by lszomoru**

### CODE REVIEW

File: src/vs/workbench/services/workspaces/common/workspaceTrust.ts [hunk 1]
```diff
@@ -129,7 +133,8 @@ export class WorkspaceTrustManagementService extends Disposable implements IWork
 		}
 
 		if (this.workspaceService.getWorkbenchState() === WorkbenchState.EMPTY) {
-			return true;
+			// Check memento, otherwise default to the setting value
+			return this._memento.isTrusted ?? this.configurationService.inspect<boolean>(WORKSPACE_TRUST_EMPTY_WORKSPACE).userValue ?? false;
```
Comment (sbatten) [inline src/vs/workbench/services/workspaces/common/workspaceTrust.ts:137]: I had to push a fix for the enabled setting as well. This will fail if we change the default so you should fallback to 
```ts
this.configurationService.inspect<boolean>(WORKSPACE_TRUST_EMPTY_WORKSPACE).defaultValue
```
if the userValue is not defined

File: src/vs/workbench/contrib/workspace/browser/workspace.contribution.ts [hunk 2]
```diff
@@ -441,6 +447,13 @@ Registry.as<IConfigurationRegistry>(ConfigurationExtensions.Configuration)
 					localize('workspace.trust.startupPrompt.once', "Ask for trust the first time an untrusted workspace is opened."),
 					localize('workspace.trust.startupPrompt.never', "Do not ask for trust when an untrusted workspace is opened."),
 				]
+			},
+			[WORKSPACE_TRUST_EMPTY_WORKSPACE]: {
+				type: 'boolean',
+				default: false,
+				included: !isWeb,
+				description: localize('workspace.trust.emptyWorkspace.description', "Controls whether or not the empty workspace is trusted within VS Code."),
```
Comment (sbatten) [inline src/vs/workbench/contrib/workspace/browser/workspace.contribution.ts:455]: I think the description should mention something like "by default" since it can be changed after the fact per empty window

File: src/vs/workbench/services/workspaces/common/workspaceTrust.ts [hunk 3]
```diff
@@ -418,4 +431,23 @@ export class WorkspaceTrustTransitionManager extends Disposable {
 	}
 }
 
+class WorkspaceTrustMemento {
```
Comment (sbatten) [inline src/vs/workbench/services/workspaces/common/workspaceTrust.ts:434]: the title and usage of this memento aren't aligned. i understand with context that this is a memento for determining the trust state of empty workspaces, but if someone else comes in they might think this memento is how we determine workspace trust for all workspaces.

File: src/vs/workbench/contrib/workspace/browser/workspace.contribution.ts [hunk 4]
```diff
@@ -441,6 +447,13 @@ Registry.as<IConfigurationRegistry>(ConfigurationExtensions.Configuration)
 					localize('workspace.trust.startupPrompt.once', "Ask for trust the first time an untrusted workspace is opened."),
 					localize('workspace.trust.startupPrompt.never', "Do not ask for trust when an untrusted workspace is opened."),
 				]
+			},
+			[WORKSPACE_TRUST_EMPTY_WORKSPACE]: {
+				type: 'boolean',
+				default: false,
```
Comment (sbatten) [inline src/vs/workbench/contrib/workspace/browser/workspace.contribution.ts:453]: one experience I was thinking about is the brand new user experience. they will open a new window and have a restricted mode banner and status bar entry. perhaps this is justification to use true as the default, or alternately, we wait until any files are opened to show the banner

File: src/vs/workbench/services/workspaces/common/workspaceTrust.ts (hunk 1 above)
Comment (lszomoru) [reply src/vs/workbench/services/workspaces/common/workspaceTrust.ts]: Given that the setting is defined with the `Application` scope, I wonder if we can use `getValue()` instead of `inspect`.

File: src/vs/workbench/services/workspaces/common/workspaceTrust.ts (hunk 3 above)
Comment (lszomoru) [reply src/vs/workbench/services/workspaces/common/workspaceTrust.ts]: I was thinking about that and for consistency I have removed the `_isWorkspaceTrusted` field and store the current workspace trust state in the memento as well. This way independent of the workspace type, the current workspace trust state is being stored consistently.

File: src/vs/workbench/services/workspaces/common/workspaceTrust.ts (hunk 1 above)
Comment (sbatten) [reply src/vs/workbench/services/workspaces/common/workspaceTrust.ts]: i think you are right, this was not possible when the setting was hidden, that's why it was this way for the enablement case.

**Review by sbatten — APPROVED** [review]

File: src/vs/workbench/contrib/workspace/browser/workspaceTrustEditor.ts [hunk 5]
```diff
@@ -783,7 +814,7 @@ export class WorkspaceTrustEditor extends EditorPane {
 	private addTrustButtonToElement(parent: HTMLElement): void {
 		const trustUris = async (uris?: URI[]) => {
 			if (!uris) {
-				await this.workspaceTrustManagementService.setWorkspaceTrust(true);
+				await this.workspaceTrustManagementService.setWorkspaceTrust(true, acceptsNonWorkspaceFiles(this.editorService, this.workspaceService));
```
Comment (sbatten) [inline src/vs/workbench/contrib/workspace/browser/workspaceTrustEditor.ts:817]: this part seems to complicate our API and I wonder if it should. when the user trusts a workspace, they may think about all the editors they have open, but they may not consider that they have already introduced untrusted files so I'm not sure we should be smart here and instead keep the API simple and prompt if additional files are introduced.

File: src/vs/workbench/services/workspaces/common/workspaceTrust.ts [hunk 6]
```diff
@@ -464,4 +497,49 @@ export class WorkspaceTrustTransitionManager extends Disposable {
...
+	set acceptsNonWorkspaceFiles(value: boolean) {
+		this._mementoObject[this._acceptsNonWorkspaceFilesKey] = value;
+		this._memento.saveMemento();
+	}
+
+	get canSetWorkspaceTrust(): boolean {
+		return this._mementoObject[this._canSetWorkspaceTrust] ?? true;
+	}
+
+	set canSetWorkspaceTrust(value: boolean) {
```
Comment (sbatten) [inline src/vs/workbench/services/workspaces/common/workspaceTrust.ts:526]: adding this was confusing to me, based on its usage, I interpret this will be for resolvers, but that's not clear to me. also, this doesn't seem to be part of this PR. can we pull this out of this change until its clear how its going to be used.

**Review by sbatten — APPROVED** [review]
With the exception of my 2 comments, this all makes sense to me.

File: src/vs/workbench/services/workspaces/common/workspaceTrust.ts (hunk 6 above)
Comment (lszomoru) [reply src/vs/workbench/services/workspaces/common/workspaceTrust.ts:526]: Removed the code.

File: src/vs/workbench/contrib/workspace/browser/workspaceTrustEditor.ts (hunk 5 above)
Comment (lszomoru) [reply src/vs/workbench/contrib/workspace/browser/workspaceTrustEditor.ts]: Agreed. Removed it.

File: src/vs/workbench/contrib/workspace/browser/workspace.contribution.ts (hunk 4 above)
Comment (lszomoru) [reply src/vs/workbench/contrib/workspace/browser/workspace.contribution.ts:479]: That is something that we can adjust based on user feedback.
